### ISSUE TITLE
MappedSchema.withInput("").withOutput("abc") prepends schema qualification to derived table references

### ISSUE THREAD

**Issue by navkast** [jOOQ/jOOQ#9814]
### Expected behavior and actual behavior:
For the following sample query, which I read using `db.parser.parse()`where `db` is an instance of `DSLContext`:
```sql
SELECT A.field2 FROM (
  SELECT field1, field2, field3 from table1
) as A
LEFT JOIN table2 
ON A.field1 = table2.field1
```
The following code adds a schema to`A`:
```java
// redirect all SQL requests that don't specify the schema to prefix with the test schema name
new MappedSchema().withInput("").withOutput("TEST_1")));
```
**Expected behavior:** aliases are left alone in remapping schemas. In jOOQ 3.12.1, aliases were left alone in remapping schemas. In this case, `A.field2` remains `A.field2`while `table2.field1` becomes `TEST_1.table2.field1`.
**Actual behavior:** In this case, `A.field2` becomes `TEST_1.A.field2` unexpectedly, and `table2.field1` becomes `TEST_1.table2.field1` as expected.

I feel like there's no valid case for adding schemas to subquery aliases? But, I don't really know.

### Versions:

- jOOQ: 3.12.4
- Java: 11
- Database (include vendor): MySQL, H2

**Comment by lukaseder** [comment]
Related: #9581

**Comment by lukaseder** [comment]
Thanks a lot for your message. This regression has probably been introduced by https://github.com/jOOQ/jOOQ/issues/8991, backported to 3.12.2 via https://github.com/jOOQ/jOOQ/issues/9299

I suspect the problem is parser specific as we are currently not implementing name resolution in the parser like we should, so when `A.field1` is parsed, the parser doesn't "know" that `A` is a table alias, not an actual schema object.

**Comment by lukaseder** [comment]
We'll need to refactor our parser a little bit on order to delay looking up table names.

Currently, we're only passing through the SQL string once, which means that while in the `WHERE` clause, we might have a lookup table available for `A`, we don't have that table yet in the `SELECT` clause. Not sure if it's easy to implement, especially because we can have cases like these:

```sql
SELECT 
  A.field2, 
  (SELECT A.field2), -- Still the outer `A`
  (SELECT A.field2 FROM table3 AS A LIMIT 1) -- A different `A`
FROM (
  SELECT field1, field2, field3 from table1
) AS A
LEFT JOIN table2 
ON A.field1 = table2.field1
```

We have the infrastructure, e.g. `org.jooq.impl.ScopeStack` to register all symbols on each scope level. Using that, we could register a `TableProxy`, similar to the existing `FieldProxy` (which has been added to delay data type lookups in the parser: #9061)

**Comment by ndiaz-addepar** [comment]
<ping>

**Comment by ndiaz-addepar** [comment]
... ping

**Comment by lukaseder** [comment]
@ndiaz-addepar Thanks for your pings. The latency in the response to your ping is explained simply by the fact that this is very hard to get right (jOOQ not necessarily knowing what a table reference is, especially if you construct it dynamically using `DSL.table(Name)`), there's a high probability that you can solve your underlying use-case very differently.

So:

1. What's the use-case of adding a schema to unqualified identifiers? Why not qualify the identifiers from the beginning?
2. What other options of implementing this use-case have been explored?

**Comment by lukaseder** [comment]
Finally looking into this. The issue wasn't parser specific but happens with ordinary queries as well, e.g. if there's no schema, or a schema is marked as the default schema, or if `TableImpl` classes are hand-written rather than generated (in case of which users probably don't specify any schema either).

The issue was indeed quite hard to fix correctly as:

- Additional tables have to be added to the scope (including derived tables). This is risky as there may be assumptions throughout jOOQ's code that all tables registered with the scope are `TableImpl` only (mainly for implicit join purposes).
- Due to the above change, we now need to be able to hide things from the current scope: https://github.com/jOOQ/jOOQ/issues/17420

The fix isn't complete yet and produced a minor regression in a more cosmetic feature:

- https://github.com/jOOQ/jOOQ/issues/17419

I think that regression is acceptable as a price to pay for fixing this rather important bug, though knowing this regression and how hard it was to fix this issue, I won't backport the fix to any minor release before 3.20.0.

**Comment by lukaseder** [comment]
I'll look into this related problem as well, soon:

- https://github.com/jOOQ/jOOQ/issues/12579
